### PR TITLE
Introduce a `Logic` class for bitwise logical operations

### PR DESCRIPTION
**Pull request by brianhuffman**
Currently in Cryptol there is no class constraint on bitwise logical operations: their types are completely polymorphic. This only makes sense because Cryptol has a closed universe of types, all of which happen to support bitwise logical operations.

Things get a bit awkward if we want to add new types to Cryptol, e.g. `Integer` (https://github.com/GaloisInc/cryptol/tree/integer). It *is* possible to define AND, XOR, etc. for unbounded integers, but SMT solvers have no support for these operations, so programs that use them won't work with `:sat` or `:prove`. It would be better if we could just exclude them. Things will get even more awkward if we decide to add a type of real numbers or user-defined uninterpreted or abstract types to Cryptol.

This pull request adds a `Logic` type class constraint to Cryptol, removing the built-in assumption of a closed universe of types. Here are all of the affected type declarations from the Cryptol prelude:
```
primitive complement : {a} (Logic a) => a -> a
primitive (&&) : {a} (Logic a) => a -> a -> a
primitive (||) : {a} (Logic a) => a -> a -> a
primitive (^) : {a} (Logic a) => a -> a -> a
primitive zero : {a} (Logic a) => a
primitive (<<) : {a, b, c} (fin b, Logic c) => [a]c -> [b] -> [a]c
primitive (>>) : {a, b, c} (fin b, Logic c) => [a]c -> [b] -> [a]c
```
Note that shift operators get a `Logic` constraint on the element type because they can shift in `zero` values.

This is technically a backward-incompatible change, as users might have to add `Logic a` constraints to their programs, but it doesn't seem to be much of an issue in practice because user code is usually isn't so polymorphic. I couldn't find any programs in the `examples` directory that required changes. I only had to add constraints to two declarations in `Cryptol::Extras`:
```
not : {a} (Logic a) => a -> a
sum : {a,n} (fin n, Logic a, Arith a) => [n]a -> a
```

Another benefit of using an explicit `Logic` constraint will be for translating Cryptol to other systems (e.g. saw-core) that have a larger universe of types. Having a `Logic` class makes it possible to use a more straightforward dictionary-passing translation, which will let us translate truly-polymorphic functions (without `Logic` constraint) onto truly polymorphic functions in the target system.

### CODE REVIEW

**Comment by brianhuffman** [comment]
A further modification we might consider is to move `zero` into its own class. For types like `Integer` or `Real` where bitwise operations are not properly supported, a `zero` value would still be useful to define things like `sum`: 
```
sum : {a,n} (fin n, Logic a, Arith a) => [n]a -> a
sum xs = foldl (+) zero xs
```
This requires `Logic a` because of the type of `zero`. But we could relax it to this:
```
sum : {a,n} (fin n, Zero a, Arith a) => [n]a -> a
```
if we changed the type of `zero`:
```
zero : {a} (Zero a) => a
```

**Comment by robdockins** [comment]
I think I'm probably on board with this.  I prefer the variant where `zero` is broken into a separate class... or maybe it should be merged into `Arith`?  We lose only `zero :: Bit` I think if `zero` became a member of `Arith`.

**Comment by brianhuffman** [comment]
I wouldn't want to put `zero` into class `Arith`: We'd lose not only `Bit` but also every tuple and record type that contains a `Bit` as a field. We would have a whole collection of types that are in fact boolean algebras, but where we don't have any way of writing the bottom element. I think it's better to have it in its own `Zero` class.

I don't know how much trouble it would be to add sub/super-classes to Cryptol, but it might be nice if class `Zero` could be made into a superclass of both `Logic` and `Arith`.

**Comment by yav** [comment]
I think this is a good idea.   @weaversa, @jpziegler what do you guys think?  It seems like it should not be a hugely disruptive change, and it will keep us more honest, which would simplify other tools, and make us more future proof.

**Comment by weaversa** [comment]
I'm worried that adding more type classes will result in more wordy type declarations.

It would be nice if there could be some assistance on this front, for example, I would like to see some work on both #373 and on the idea of providing _some_ of the type and letting cryptol fill in the rest, ex.)

```
sum : {a,n} (fin n, ...) => [n]a -> a
sum xs = foldl (+) zero xs
```

**Comment by jpziegler** [comment]
I'm concerned about the wordiness as well. And I can imagine happily writing Cryptol code as I have always done, not bothering with the `Logic` keyword, then suddenly I bitwise-and two bitvectors together and it gives me that annoying "Logic" error message.

I agree that allowing type completion (the `...` that @weaversa mentions) would do a lot to mitigate this. Is there a ticket for that?

I can think of a few other approaches, each unsatisfactory in its own way:

1. Add an `Unconstrained` class that includes Integers and the like. But by default, everything is `Logic` unless specified otherwise.
2. Add some kind of `allow-integers` mode (and similar). With none of these modes enabled, everything is assumed to be in `Logic`.

I also found the following comment interesting:

> It _is_ possible to define AND, XOR, etc. for unbounded integers, but SMT solvers have no support for these operations, so programs that use them won't work with `:sat` or `:prove`.

This comment implies that the design of Cryptol is driven by the capabilities of SMT, rather than what we would like to have in the Cryptol language itself. I think it would be a fair argument to say that Cryptol (and other tools that use SMT) should drive SMT development, not the other way around!

That said, I don't think I really want to bitwise-and my unbounded integers.

In addition, as we move towards post-quantum, I see requirements for things like floating-point in Cryptol's future for which bitwise operations really don't make sense.

In summary, I agree that this is a problem that we must address sooner or later, but I have concerns about usability and readability.

**Comment by yav** [comment]
I wouldn't expect that you'd have to write many more constraints.   As @brianhuffman points out, this change did not require any new constraints to be added in any of the examples bundled with Cryptol.

I suspect that writing `Logic` will be about as common as writing `Arith`, and I can't recall a case in recent history when I've had to provide an explicit `Arith` constraint when writing Cryptol.   Is this something that happens a lot in your code base?

As for #373, this is very much on the table, we just have to find some time to implement it.   The `...` form is also an option, but that one requires some more thinking to figure out exactly how it should work.

**Comment by brianhuffman** [comment]
@jpziegler writes:
> suddenly I bitwise-and two bitvectors together and it gives me that annoying "Logic" error message.

I should point out that if you're doing anything with *bitvectors*, then you'll never introduce any new `Logic`, `Arith` or `Cmp` constraints because the types are specific enough that all such constraints are discharged automatically.

The only time you will ever need a `Logic` constraint is if you do bitwise operations on a completely unknown type like `a`. It is quite uncommon for code to be written at this level of generality unless you're writing a general-purpose library like `Cryptol::Extras`.

**Comment by weaversa** [comment]
> The only time you will ever need a Logic constraint is if you do bitwise operations on a completely unknown type like a. It is quite uncommon for code to be written at this level of generality unless you're writing a general-purpose library like Cryptol::Extras.

But, what if you have a perfectly wonderful function that Cryptol discharges without a user supplied type and then you want to add a single constraint, say `x <= 256` --- now you are _forced_ to add all of these extra constraints/type classes. This is already happening w/ `Arith` and `Comp` and can be quite annoying.

**Comment by weaversa** [comment]
It upsets me enough that I have to add `fin` to all my variables. Luckily I found a trick which is `fin (a+b+c+d+e)` etc.

```
test : {a, b, c, d, e, f, g} (fin (a+b+c+d+e+f+g)) => ([a], [b], [c], [d], [e], [f], [g]) -> [a]
test (a, b, c, d, e, f, g) = a
```
